### PR TITLE
Multi step 90

### LINKED ISSUE #90: Multi-step simulator function
Created from #70

> we may want to have the simulator batch core access updates so that we don't get 80000 one by one for performance reasons.

> Currently, the UI is controlling the number of steps per second by repeatedly calling step within setInterval. It might be nice to allow a number of steps to be passed to the step function (defaulting to 1). If multiple steps are requested it executes them all as quickly as possible, generating the minimum number of required messages.

> However,  I think this is probably one for the future as I'm not sure how much performance gain this would really achieve...

### PR DESCRIPTION
**Pull request by gareththegeek**
Resolves #90 - only publishes the necessary number of messages and allows a number of steps to be specified when calling `step` method.  If no steps specified the `simulator` will advance a single step.

### CODE REVIEW

**Comment by dougajmcdonald** [comment]
In testing there are a few events which give me their info in the format:

```
{
  type: <int>,
  payload: {}
}
```

The contents of payload seem ok, just not unwrapped. Affected events which I've noticed.

- ROUND_ENDED
- RUN_PROGRESS
- CORE_ACCESS

Happy to use `data.payload` on the UI (as I have done at the mo) if it was intended. If not, fix that shit up on this branch and push it real good

**Comment by dougajmcdonald** [comment]
Oh BTW, other than that, this PR combined with my removal of `setInterval` on the client has crushed the performance to new unseen levels!

File: index.ts [hunk 1]
```diff
@@ -1,70 +1,72 @@
...
-import { Publisher } from "./simulator/Publisher"
-import { IOptions } from "./simulator/interface/IOptions"
-import { ILoader } from "./simulator/interface/ILoader"
-import { IState } from "./simulator/interface/IState"
-import { IInstruction } from "./simulator/interface/IInstruction"
-import { IPublishProvider } from "./simulator/interface/IPublishProvider"
-
-import * as clone from "clone"
-import { ICoreLocation } from "./simulator/interface/ICoreLocation"
+﻿import { IParser } from "./parser/interface/IParser";
```
Comment (dougajmcdonald) [inline index.ts]: LOL, animal!

File: index.ts [hunk 2]
```diff
@@ -103,40 +112,46 @@ class Api {
...
-        return clone(this.core.getWithInfoAt(address))
+
+        return clone(this.core.getWithInfoAt(address));
     }
 
-    public step() : void {
-        this.simulator.step()
+    public step(steps?: number) : void {
+
+        this.simulator.step(steps);
```
Comment (dougajmcdonald) [inline index.ts]: This should probably be `: number` and `return this.simulator.step(steps)` to mirror the internal API. I'm not using the retVal atm, but just for consistency sake

File: simulator/Publisher.ts [hunk 3]
```diff
@@ -17,42 +17,39 @@ export class Publisher implements IPublisher {
...
+            .forEach(s => {
+                var message = s.dequeue();
+
+                if (!message) {
+                    return;
+                }
 
-        this.typeEnabled[type] = enabled;
+                this.publishProvider.publishSync(
+                    this.typeDictionary[message.type],
```
Comment (dougajmcdonald) [inline simulator/Publisher.ts]: Should this just be `message`?

Comment (gareththegeek) [reply simulator/Publisher.ts]: The `typeDictionary` bit turns enums into strings e.g. 1 becomes CORE_ACCESS etc.

**Pull request merged**